megamek/src/megamek/client/bot/princess/FireControl.java

```diff
@@ -2836,4 +2823,38 @@ protected Mounted getAntiAirAmmo(List<Mounted> ammoList, WeaponType weaponType,
 
         return returnAmmo;
     }
+
+    // Helper method that figures out the valid facing changes for the given shooter
+    private ArrayList<Integer> getValidFacingChanges(Entity shooter)
+    {
+    	// figure out all valid twists or turret turns
+        // mechs can turn:
+        //		one left, one right unless he has "no torso twist" quirk
+        //		two left, two right if he has "extended torso twist" quirk
+        // vehicles and turrets can turn any direction unless he has no turret
+        ArrayList<Integer> validFacingChanges = new ArrayList<Integer>();
+        if(shooter instanceof Mech &&
+        	!shooter.hasQuirk(OptionsConstants.QUIRK_NEG_NO_TWIST))
+        {
+        	validFacingChanges.add(1);
+            validFacingChanges.add(-1);
+
+        	if(shooter.hasQuirk(OptionsConstants.QUIRK_POS_EXT_TWIST))
+        	{
+                validFacingChanges.add(2);
+                validFacingChanges.add(-2);
+        	}
+        }
+        else if(shooter instanceof Tank &&
+        		!((Tank) shooter).hasNoTurret())
+        {
+        	validFacingChanges.add(1);
+        	validFacingChanges.add(-1);
+        	validFacingChanges.add(2);
+        	validFacingChanges.add(-2);
+        	validFacingChanges.add(3);
+        }
+        
+        return validFacingChanges;
+    }
```

Comment:
Overall, seems like there's a lot of duplicate code here (which is true of the MM project in general) that could be moved into helper methods.